lib/recursive-methods.rb

```diff
@@ -1,49 +1,94 @@
+require "pry"
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
+# Space complexity: O(n) where the stack is n levels deep
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  if (n == 1) || (n == 0)
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
```

Comment:
Not quite since s[0..-1] makes a new string, it's O(n^2)

It's similar for Space complexity.